src/core/fetchers/segment/promise_prioritizer.ts

```diff
@@ -0,0 +1,321 @@
+import log from "../../../log";
+import arrayFindIndex from "../../../utils/array_find_index";
+import TaskCanceller from "../../../utils/task_canceller";
+
+export default class PromisePrioritizer<T> {
+  private _minPendingPriority: number | null;
+  private _waitingQueue: Array<IPrioritizerTask<T>>;
+  private _pendingTasks: Array<IPrioritizerTask<T>>;
+  private _prioritySteps: IPrioritizerPrioritySteps;
+
+  constructor({ prioritySteps }: IPrioritizerOptions) {
+    this._minPendingPriority = null;
+    this._waitingQueue = [];
+    this._pendingTasks = [];
+    this._prioritySteps = prioritySteps;
+
+    if (this._prioritySteps.high >= this._prioritySteps.low) {
+      throw new Error(
+        "FP Error: the max high level priority should be given a lower" +
+          "priority number than the min low priority."
+      );
+    }
+  }
+
+  public create(
+    promise: () => Promise<T>,
+    priority: number,
+    canceller: TaskCanceller
+  ) {
+    let newTask: IPrioritizerTask<T>;
+    return new Promise<T>((resolve, reject) => {
+      const trigger = (shouldRun: boolean) => {
+        if (!shouldRun) return;
+        this._minPendingPriority =
+          this._minPendingPriority === null
+            ? newTask.priority
+            : Math.min(this._minPendingPriority, newTask.priority);
+        this._pendingTasks.push(newTask);
+        newTask
+          .promise()
+          .then(resolve)
+          .catch(reject)
```

Comment:
This promise (`newTask.promise`) may also reject when we're temporarily cancelling it, no?
In which case we might not want to call `reject`